genesyscloud/architect_ivr/genesyscloud_architect_ivr_proxy.go

```diff
@@ -306,3 +308,81 @@ func (a *architectIvrProxy) uploadDnisChunk(ctx context.Context, ivr platformcli
 	}
 	return putIvr, resp, nil
 }
+
+// getIvrDnisAndChunks returns the dnis array to attach to the ivr on the initial POST/PUT
+// along with the chunks to PUT after, if necessary.
+func (a *architectIvrProxy) getIvrDnisAndChunks(ctx context.Context, id string, post bool, ivr *platformclientv2.Ivr) (*[]string, [][]string, error) {
+	var (
+		dnisChunks                [][]string
+		dnisSliceForInitialUpload *[]string
+	)
+
+	// Create
+	if post {
+		dnisChunks = utillists.ChunkStringSlice(*ivr.Dnis, a.maxDnisPerRequest)
+		dnisSliceForInitialUpload = &dnisChunks[0]
+		dnisChunks = dnisChunks[1:] // all chunks after index 0, if they exist
+		return dnisSliceForInitialUpload, dnisChunks, nil
+	}
+
+	// Update
+	// read the ivr to get current dnis array
+	currentIvr, _, err := a.getArchitectIvr(ctx, id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// slice to establish what we're adding
+	dnisToAdd := utillists.SliceDifference(*ivr.Dnis, *currentIvr.Dnis)
+
+	// chunk that if necessary
+	if len(dnisToAdd) > a.maxDnisPerRequest {
+		dnisChunks = utillists.ChunkStringSlice(dnisToAdd, a.maxDnisPerRequest)
+		dnisForInitialCall := removeItemsToBeAddedFromOriginalList(*ivr.Dnis, dnisToAdd)
+		// append the first chunk
+		dnisForInitialCall = append(dnisForInitialCall, dnisChunks[0]...)
+		// return dnis for initial upload along with any chunks that may exist after index 0
+		return &dnisForInitialCall, dnisChunks[1:], nil
+	}
+
+	// no chunking logic necessary
+	return ivr.Dnis, nil, nil
+}
+
+// removeItemsToBeAddedFromOriginalList is used to remove the new did numbers from the dnis slice that we collected from the schema
+// We do this because we are keeping the new numbers separate in the chunks slice to be uploaded subsequently
+func removeItemsToBeAddedFromOriginalList(allDnis []string, toBeAdded []string) []string {
+	for _, number := range toBeAdded {
+		allDnis = utillists.Remove(allDnis, number)
+	}
+	return allDnis
+}
+
+// resetArchitectIvrAfterFailedChunkUpload
+// If the chunking logic failed on create - delete the IVR that was created
+// Otherwise the CreateContext func will fail and terraform will assume the IVR was never created
+// If it failed on update - reset the ivr to its original state. This is done so that subsequent applies
+// will output the same plan as before the last update failed
+func (a *architectIvrProxy) resetArchitectIvrAfterFailedChunkUpload(ctx context.Context, post bool, id string, originalConfig *platformclientv2.Ivr) error {
+	if post {
+		log.Printf("Deleting IVR %s because chunking logic failed on create", id)
+		if _, err := a.deleteArchitectIvr(ctx, id); err != nil {
+			return fmt.Errorf("failed to delete ivr %s after dnis chunking logic failed: %v", id, err)
+		}
+		log.Printf("Deleted IVR %s", id)
+		return nil
+	}
+
+	if originalConfig == nil {
+		return fmt.Errorf("cannot reset IVR %s without a copy of the original configuration", id)
+	}
+
+	originalConfig.Version = nil
```

Comment:
why is this assigned to nil

Reply:
The object originalConfig was read using a get earlier in the process, before a couple update happens and the value of version was incremented a few times, so trying to do a put with an old version number was causing weird behaviour. Better to just omit it and let the API compute it